### ISSUE TITLE
Calling queue_free() on many intersecting Area2D nodes causes crash since 3.5

### ISSUE THREAD

**Issue by warodot** [godotengine/godot#67273]
### Godot version

3.5.1 stable

### System information

Windows 10, GLES 3, NVIDIA GeForce GTX 1060 3GB

### Issue description

Calling queue_free() on many intersecting Area2D nodes (around ~50 or more, apparently depends on hardware) may crash without any errors reported on console if done in a single frame, more Area2Ds seem to cause the crash more often.

Also tested on **Godot 3.5** with no difference, but when tested on **Godot 3.4.4** it didn't seem to crash regardless of intersecting Area2Ds. **Godot 3.5.1** on MacOS also appears to not have this issue.

### Steps to reproduce

1- Have many Area2Ds intersecting each other
2- Call queue_free() on them in a single frame

**How to use reproduction project:**
1- Run reproduction project
2- Press "Enter", the main node will delete all of its area2D children (100 of them) at once, causing the crash

It may be necessary to repeat a few times to reproduce the crash, adding more Area2D nodes causes it to crash more often.

### Minimal reproduction project

[Area2DCrash.zip](https://github.com/godotengine/godot/files/9759198/Area2DCrash.zip)

**Comment by cbinners** [comment]
Also confirming that this issue is still present.

**Comment by Calinou** [comment]
@warodot @viksl @cbinners Can you reproduce this in any of the [3.5 betas and RCs](https://downloads.tuxfamily.org/godotengine/3.5/) to determine when the regression started?

**Comment by cbinners** [comment]
@Calinou I'll do some testing and will report back shortly

**Comment by cbinners** [comment]
So far I've been able to determine it broke between 3.5beta1 and 3.5beta2 releases. Going to attempt to bisect now

**Comment by cbinners** [comment]
I haven't extensively tested it, but reverted this commit fixes my local issues.

**Comment by akien-mga** [comment]
CC @godotengine/physics

**Comment by lawnjelly** [comment]
I can't seem to reproduce this on my system (Linux Mint 20), despite running lots of times with 3.5 and 3.5.1, and using both the original project and then adding higher numbers of Areas like 300 etc. It could just be my system but has anyone reproduced this on a non-windows OS?

**Comment by cbinners** [comment]
I haven't had the chance to test it on a non-windows build.

I was able to reproduce it 100% of the time with 500 areas on my system (Windows 10, Ryzen 5950x, Titan RTX, GLES 3)

**Comment by kleonc** [comment]
For me it hasn't crashed for 400 areas, for 500 it crashed. Crash log from a build from week ago (v3.6.beta.custom_build [01574a566]):
```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v3.6.beta.custom_build (01574a566ff1f1a7270bd986b226c1bd351f894e)
Dumping the backtrace. Please include this when reporting the bug on: https://github.com/godotengine/godot/issues
[0] RID_Data::get_id (godot\core\rid.h:54)
[1] RID::get_id (godot\core\rid.h:84)
[2] RID::operator< (godot\core\rid.h:71)
[3] Area2DSW::BodyKey::operator< (godot\servers\physics_2d\area_2d_sw.h:80)
[4] Comparator<Area2DSW::BodyKey>::operator() (godot\core\typedefs.h:327)
[5] Map<Area2DSW::BodyKey,Area2DSW::BodyState,Comparator<Area2DSW::BodyKey>,DefaultAllocator>::_find (godot\core\map.h:228)
[6] Map<Area2DSW::BodyKey,Area2DSW::BodyState,Comparator<Area2DSW::BodyKey>,DefaultAllocator>::find (godot\core\map.h:522)
[7] Map<Area2DSW::BodyKey,Area2DSW::BodyState,Comparator<Area2DSW::BodyKey>,DefaultAllocator>::operator[] (godot\core\map.h:595)
[8] Area2DSW::remove_area_from_query (godot\servers\physics_2d\area_2d_sw.h:192)
[9] Area2Pair2DSW::~Area2Pair2DSW (godot\servers\physics_2d\area_pair_2d_sw.cpp:153)
[10] Area2Pair2DSW::`scalar deleting destructor'
[11] memdelete<Constraint2DSW> (godot\core\os\memory.h:118)
[12] Space2DSW::_broadphase_unpair (godot\servers\physics_2d\space_2d_sw.cpp:1255)
[13] BroadPhase2DBVH::_unpair_callback (godot\servers\physics_2d\broad_phase_2d_bvh.cpp:99)
[14] BVH_Manager<CollisionObject2DSW,2,1,128,BroadPhase2DBVH::UserPairTestFunction<CollisionObject2DSW>,BroadPhase2DBVH::UserCullTestFunction<CollisionObject2DSW>,Rect2,Vector2,1>::_unpair (godot\core\math\bvh.h:517)
[15] BVH_Manager<CollisionObject2DSW,2,1,128,BroadPhase2DBVH::UserPairTestFunction<CollisionObject2DSW>,BroadPhase2DBVH::UserCullTestFunction<CollisionObject2DSW>,Rect2,Vector2,1>::_remove_pairs_containing (godot\core\math\bvh.h:648)
[16] BVH_Manager<CollisionObject2DSW,2,1,128,BroadPhase2DBVH::UserPairTestFunction<CollisionObject2DSW>,BroadPhase2DBVH::UserCullTestFunction<CollisionObject2DSW>,Rect2,Vector2,1>::_remove_changed_item (godot\core\math\bvh.h:741)
[17] BVH_Manager<CollisionObject2DSW,2,1,128,BroadPhase2DBVH::UserPairTestFunction<CollisionObject2DSW>,BroadPhase2DBVH::UserCullTestFunction<CollisionObject2DSW>,Rect2,Vector2,1>::erase (godot\core\math\bvh.h:219)
[18] BVH_Manager<CollisionObject2DSW,2,1,128,BroadPhase2DBVH::UserPairTestFunction<CollisionObject2DSW>,BroadPhase2DBVH::UserCullTestFunction<CollisionObject2DSW>,Rect2,Vector2,1>::erase (godot\core\math\bvh.h:153)
[19] BroadPhase2DBVH::remove (godot\servers\physics_2d\broad_phase_2d_bvh.cpp:58)
[20] CollisionObject2DSW::_set_space (godot\servers\physics_2d\collision_object_2d_sw.cpp:242)
[21] Area2DSW::set_space (godot\servers\physics_2d\area_2d_sw.cpp:77)
[22] Physics2DServerSW::area_set_space (godot\servers\physics_2d\physics_2d_server_sw.cpp:304)
[23] Physics2DServerWrapMT::area_set_space (godot\servers\physics_2d\physics_2d_server_wrap_mt.h:131)
[24] CollisionObject2D::_notification (godot\scene\2d\collision_object_2d.cpp:89)
[25] CollisionObject2D::_notificationv (godot\scene\2d\collision_object_2d.h:38)
[26] Area2D::_notificationv (godot\scene\2d\area_2d.h:38)
[27] Object::notification (godot\core\object.cpp:929)
[28] Node::_propagate_exit_tree (godot\scene\main\node.cpp:367)
[29] Node::_set_tree (godot\scene\main\node.cpp:2822)
[30] Node::remove_child (godot\scene\main\node.cpp:1346)
[31] Node::_notification (godot\scene\main\node.cpp:169)
[32] Node::_notificationv (godot\scene\main\node.h:47)
[33] CanvasItem::_notificationv (godot\scene\2d\canvas_item.h:163)
[34] Node2D::_notificationv (godot\scene\2d\node_2d.h:37)
[35] CollisionObject2D::_notificationv (godot\scene\2d\collision_object_2d.h:38)
[36] Area2D::_notificationv (godot\scene\2d\area_2d.h:38)
[37] Object::notification (godot\core\object.cpp:929)
[38] Object::_predelete (godot\core\object.cpp:388)
[39] predelete_handler (godot\core\object.cpp:2004)
[40] memdelete<Object> (godot\core\os\memory.h:111)
[41] SceneTree::_flush_delete_queue (godot\scene\main\scene_tree.cpp:1253)
[42] SceneTree::idle (godot\scene\main\scene_tree.cpp:657)
[43] Main::iteration (godot\main\main.cpp:2340)
[44] OS_Windows::run (godot\platform\windows\os_windows.cpp:3591)
[45] widechar_main (godot\platform\windows\godot_windows.cpp:176)
[46] _main (godot\platform\windows\godot_windows.cpp:198)
[47] main (godot\platform\windows\godot_windows.cpp:210)
[48] __scrt_common_main_seh (d:\A01\_work\6\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[49] BaseThreadInitThunk
-- END OF BACKTRACE --
================================================================
Segmentation fault
```

**Comment by lawnjelly** [comment]
From the callstack this is a dangling RID error. There is a RID being used in the physics that points to an object which has been deleted.

And the reason it didn't crash before the PR changing the < operator is probably because before it simply compared the address of `_data`, whereas after the PR it looks up the object at `_data` and reads the ID. If the object is deleted it will often crash, whereas before the error was masked.

**Comment by akien-mga** [comment]
Fixed by #69946.
